llvm/lib/Linker/IRMover.cpp

```diff
@@ -88,45 +69,18 @@ class TypeMapTy : public ValueMapTypeRemapper {
 private:
   Type *remapType(Type *SrcTy) override { return get(SrcTy); }
 
-  bool areTypesIsomorphic(Type *DstTy, Type *SrcTy);
+  bool tryToEstablishTypeIsomorphism(Type *DstTy, Type *SrcTy);
 };
 }
 
 void TypeMapTy::addTypeMapping(Type *DstTy, Type *SrcTy) {
-  assert(SpeculativeTypes.empty());
-  assert(SpeculativeDstOpaqueTypes.empty());
-
-  // Check to see if these types are recursively isomorphic and establish a
-  // mapping between them if so.
-  if (!areTypesIsomorphic(DstTy, SrcTy)) {
-    // Oops, they aren't isomorphic.  Just discard this request by rolling out
-    // any speculative mappings we've established.
-    for (Type *Ty : SpeculativeTypes)
-      MappedTypes.erase(Ty);
-
-    SrcDefinitionsToResolve.resize(SrcDefinitionsToResolve.size() -
-                                   SpeculativeDstOpaqueTypes.size());
-    for (StructType *Ty : SpeculativeDstOpaqueTypes)
-      DstResolvedOpaqueTypes.erase(Ty);
-  } else {
-    // SrcTy and DstTy are recursively ismorphic. We clear names of SrcTy
-    // and all its descendants to lower amount of renaming in LLVM context
-    // Renaming occurs because we load all source modules to the same context
-    // and declaration with existing name gets renamed (i.e Foo -> Foo.42).
-    // As a result we may get several different types in the destination
-    // module, which are in fact the same.
-    for (Type *Ty : SpeculativeTypes)
-      if (auto *STy = dyn_cast<StructType>(Ty))
-        if (STy->hasName())
-          STy->setName("");
-  }
-  SpeculativeTypes.clear();
-  SpeculativeDstOpaqueTypes.clear();
+  tryToEstablishTypeIsomorphism(DstTy, SrcTy);
```

Comment:
Doesn't it sound like type mapping insertion and clear name should be up to addTypeMapping, and just check if types are isomorphic? Or, at least, it feels like now the wrapper could be `tryAddTypeMapping` (and the function itself maybe `addIfTypesAreIsomorphic`).

Reply:
> Doesn't it sound like type mapping insertion and clear name should be up to addTypeMapping, and just check if types are isomorphic?

I don't think so, as that would only add the mapping for the top-level type, without also adding it for all the recursively isomorphic ones.

Reply:
Right, what about something along the lines of `RecursivelyAddMappingIfTypesAreIsomorphic`? I'm not completely sure `tryToEstablishTypeIsomorphism` is emblematic enough to convey the mapping add (though this is mentioned in the doc-comment). Not strong on this, up to you.